From c21901b3af541397c13395d7362298fc87bb9419 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Mon, 5 Dec 2022 12:06:05 +0100
Subject: [PATCH 0391/1695] drivers: stm32_rng: add health test and noise
 source control conf

Adds health test and noise source control customized configuration
according to applied RNG NIST configuration.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git f376d58b559f009321457dcea65b038d36500c3f drivers: stm32_rng: add health test and noise source control conf]
Change-Id: I033d00d6d1f37e53cebc464ccdc79c77816ad83d
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/297212
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Lionel DEBIEVE <lionel.debieve@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/stm32_rng.c | 58 +++++++++++++++++++++++++++++++++++-----
 1 file changed, 51 insertions(+), 7 deletions(-)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index ebe7c833f..c88d41151 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -26,6 +26,8 @@
 #define RNG_CR			U(0x00)
 #define RNG_SR			U(0x04)
 #define RNG_DR			U(0x08)
+#define RNG_NSCR		U(0x0C)
+#define RNG_HTCR		U(0x10)
 
 #define RNG_CR_RNGEN		BIT(2)
 #define RNG_CR_IE		BIT(3)
@@ -50,8 +52,14 @@
 #define RNG_FIFO_BYTE_DEPTH	U(16)
 
 #define RNG_NIST_CONFIG_A	U(0x0F00F00)
+#define RNG_HEALTH_CONFIG_A	U(0xAB39)
+#define RNG_NOISE_CTRL_CONFIG_A	U(0x0492)
 #define RNG_NIST_CONFIG_B	U(0x1801000)
+#define RNG_HEALTH_CONFIG_B	U(0xAAC7)
+#define RNG_NOISE_CTRL_CONFIG_B	U(0x3FFFF)
 #define RNG_NIST_CONFIG_C	U(0x0F00D00)
+#define RNG_HEALTH_CONFIG_C	U(0xAAC7)
+#define RNG_NOISE_CTRL_CONFIG_C	U(0x3FFFF)
 #define RNG_NIST_CONFIG_MASK	GENMASK_32(25, 8)
 
 #define RNG_MAX_NOISE_CLK_FREQ	U(3000000)
@@ -68,6 +76,10 @@ struct stm32_rng_instance {
 	unsigned int lock;
 	uint64_t error_to_ref;
 	uint32_t pm_cr;
+	uint32_t pm_health;
+	uint32_t pm_noise_ctrl;
+	uint32_t health_test_conf;
+	uint32_t noise_ctrl_conf;
 	uint32_t rng_config;
 	bool release_post_boot;
 	bool clock_error;
@@ -282,6 +294,16 @@ static TEE_Result init_rng(void)
 		io_clrsetbits32(rng_base + RNG_CR, RNG_CR_CLKDIV,
 				clock_div << RNG_CR_CLKDIV_SHIFT);
 
+		/*
+		 * Write health test and noise source control configuration
+		 * according to current RNG entropy source configuration
+		 */
+		if (stm32_rng->noise_ctrl_conf)
+			io_write32(rng_base + RNG_NSCR, stm32_rng->noise_ctrl_conf);
+
+		if(stm32_rng->health_test_conf)
+			io_write32(rng_base + RNG_HTCR, stm32_rng->health_test_conf);
+
 		/* No need to wait for RNG_CR_CONDRST toggle as we enable clk */
 		io_clrsetbits32(rng_base + RNG_CR, RNG_CR_CONDRST,
 				RNG_CR_RNGEN);
@@ -393,7 +415,7 @@ void plat_rng_init(void)
 }
 #endif
 
-static TEE_Result stm32_rng_pm_resume(uint32_t pm_cr)
+static TEE_Result stm32_rng_pm_resume(void)
 {
 	vaddr_t base = get_base();
 
@@ -406,21 +428,35 @@ static TEE_Result stm32_rng_pm_resume(uint32_t pm_cr)
 		 * access that set RNG_CR_CONDRST bit. Else config setting is
 		 * not taken into account.
 		 */
-		io_write32(base + RNG_CR, pm_cr | RNG_CR_CONDRST);
+		io_write32(base + RNG_CR, stm32_rng->pm_cr | RNG_CR_CONDRST);
+
+		/* Restore health test and noise control configuration */
+		io_write32(base + RNG_NSCR, stm32_rng->pm_noise_ctrl);
+		io_write32(base + RNG_HTCR, stm32_rng->pm_health);
 
 		io_clrsetbits32(base + RNG_CR, RNG_CR_CONDRST, RNG_CR_RNGEN);
 	} else {
-		io_write32(base + RNG_CR, RNG_CR_RNGEN | pm_cr);
+		io_write32(base + RNG_CR, RNG_CR_RNGEN | stm32_rng->pm_cr);
 	}
 
 	return TEE_SUCCESS;
 }
 
+static TEE_Result stm32_rng_pm_suspend(void)
+{
+	vaddr_t base = get_base();
+
+	stm32_rng->pm_cr = io_read32(base + RNG_CR);
+	stm32_rng->pm_health = io_read32(base + RNG_HTCR);
+	stm32_rng->pm_noise_ctrl = io_read32(base + RNG_NSCR);
+
+	return TEE_SUCCESS;
+}
+
 static TEE_Result
 stm32_rng_pm(enum pm_op op, unsigned int pm_hint __unused,
 	     const struct pm_callback_handle *pm_handle __unused)
 {
-	static uint32_t pm_cr;
 	TEE_Result res = TEE_ERROR_GENERIC;
 
 	assert(stm32_rng && (op == PM_OP_SUSPEND || op == PM_OP_RESUME));
@@ -429,10 +465,10 @@ stm32_rng_pm(enum pm_op op, unsigned int pm_hint __unused,
 	if (res)
 		return res;
 
-	if (op == PM_OP_SUSPEND)
-		pm_cr = io_read32(get_base() + RNG_CR);
+	if (op == PM_OP_RESUME)
+		res = stm32_rng_pm_resume();
 	else
-		res = stm32_rng_pm_resume(pm_cr);
+		res = stm32_rng_pm_suspend();
 
 	clk_disable(stm32_rng->clock);
 
@@ -478,6 +514,8 @@ static TEE_Result stm32_rng_parse_fdt(const void *fdt, int node)
 	if (!cuint) {
 		IMSG("No RNG configuration specified, defaulting to NIST B");
 		stm32_rng->rng_config = RNG_NIST_CONFIG_B;
+		stm32_rng->health_test_conf = RNG_HEALTH_CONFIG_B;
+		stm32_rng->noise_ctrl_conf = RNG_NOISE_CTRL_CONFIG_B;
 		goto end;
 	}
 
@@ -485,12 +523,18 @@ static TEE_Result stm32_rng_parse_fdt(const void *fdt, int node)
 	switch (config_id) {
 	case 0:
 		stm32_rng->rng_config = RNG_NIST_CONFIG_A;
+		stm32_rng->health_test_conf = RNG_HEALTH_CONFIG_A;
+		stm32_rng->noise_ctrl_conf = RNG_NOISE_CTRL_CONFIG_A;
 		break;
 	case 1:
 		stm32_rng->rng_config = RNG_NIST_CONFIG_B;
+		stm32_rng->health_test_conf = RNG_HEALTH_CONFIG_B;
+		stm32_rng->noise_ctrl_conf = RNG_NOISE_CTRL_CONFIG_B;
 		break;
 	case 2:
 		stm32_rng->rng_config = RNG_NIST_CONFIG_C;
+		stm32_rng->health_test_conf = RNG_HEALTH_CONFIG_C;
+		stm32_rng->noise_ctrl_conf = RNG_NOISE_CTRL_CONFIG_C;
 		break;
 	default:
 		panic("RNG config not supported");
-- 
2.47.2

